### PR TITLE
Add lines changed for CR column

### LINKED ISSUE #225: RFC: Pulldasher: Show number of lines changed on PRs
@andyg0808 commented on [Tue Oct 12 2021](https://github.com/iFixit/ifixit/issues/40208)

## Problem
We don't have a quick way to see which PRs are likely to be easiest to CR.

## Do This
Add an indicator for number of lines changed to Pulldasher.

## Background
Nerd Herd's brought this up twice in retros now.

CC: @Annika29 

---

@davidrans commented on [Tue Oct 12 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-941294293)

Some discussion happened here as well: https://ifixit.slack.com/archives/C90126AN4/p1628184142080500

---

@djmetzle commented on [Tue Oct 12 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-941308854)

If the intent is to use LoC changed as a proxy for complexity, lets craft some arbitrary metric like:
```
log_10(lines_added) + log_2(lines_deleted)
```

Or the likes? That way we're not encouraging gaming LoC, just discouraging very large PRs?

---

@davidrans commented on [Tue Oct 12 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-941458336)

How about median LoC per commit?

---

@GrantMatejka commented on [Wed Oct 13 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-942417911)

I like the idea of this feature, and even though LoC is a notoriously bad metric, I think if we use it _innocently_ and bucket<sup>1</sup> on some convoluted calculation it would work fine. I think an issue like this is very open to bikeshedding so fwmioiw I'm fine trying anything to see if it works/is helpful

<sup>1</sup>bucket as in a pull is either 'long' or not, so on pulldasher maybe a long pulls title is yellow or something but all 'short' pulls don't have any indicator idk

---

@djmetzle commented on [Fri Oct 29 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-954918372)

Should we migrate this to the pulldasher repo? Or was the intent here to gather consensus? 

---

@davidrans commented on [Fri Oct 29 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-954932738)

Copying my comment from the slack thread here:

I like it. I've been opposed to it in the past, cause I didn't want people to skip long pulls, but if you don't have time to CR a super long pull it's nice to know before you click. And I also think it'll be cool if it results in smaller pulls getting CR'd quicker. It's frustrating when you work hard to break something up into small, readable pulls and it still takes a long time to get CR.

I already added lines changed, and # of commits to the pulldasher DB so it should be pretty easy

---

@danielbeardsley commented on [Fri Oct 29 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-954937542)

> I already added lines changed, and # of commits to the pulldasher DB so it should be pretty easy

Oh nice!

---

@andyg0808 commented on [Mon Nov 01 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-956570443)

@djmetzle Probably reasonable to migrate to the Pulldasher repo.

---

@djmetzle commented on [Tue Nov 02 2021](https://github.com/iFixit/ifixit/issues/40208#issuecomment-958174937)

@andyg0808 
> @djmetzle Probably reasonable to migrate to the Pulldasher repo.

image

Doesn't seem to want to let me. Maybe someone else can give it a try?

### PR DESCRIPTION
**Pull request by jordycosta**
### Preview:

Screenshot 2023-03-23 134106 night

---

### Changes Made:

1. Included additions/deletions fields in PullData
2. Added a new pull card
3. Added a column property
4. Set a conditional render

### CR Notes:

Review to https://github.com/iFixit/pulldasher/pull/365/commits/c4cedb535a9d83a3fd919608335f6ef40ea87cff

---

Closes #225

### CODE REVIEW

File: frontend/src/pull-card/index.tsx [hunk 1]
```diff
@@ -109,6 +109,69 @@ export const PullCard = memo(function PullCard({
   );
 });
 
+export const PullCardWithLinesChanged = memo(function PullCard({
```
[the remark this answers is not in the document]
Comment (jordycosta) [reply frontend/src/pull-card/index.tsx]: Good idea, I updated this via https://github.com/iFixit/pulldasher/pull/365/commits/a381dca2d88f1189e5f7be0dbee374753efe2195 to https://github.com/iFixit/pulldasher/pull/365/commits/f433975f8dea2743c25312c62a89329d22b5dfa1

File: frontend/test/pull-card-demo.tsx [hunk 2]
```diff
@@ -66,7 +66,7 @@ function Row({ title, pullDatas }: { title: string; pullDatas: PullData[] }) {
               border="1px solid var(--panel-default-border)"
               overflow="hidden"
             >
-              <PullCard pull={new Pull(pullData)} show={true} />
```
Comment (danielbeardsley) [inline frontend/test/pull-card-demo.tsx:69]: Can we add a new example to the pull-card-demo that shows these values? Generally, that's the page I use while developing the `PullCard` component so I can see how my changes affect all the different UI states a card can take.

**Comment by danielbeardsley** [comment]
> Could we consider adding a CR leaderboard for lines reviewed

If we do that, I think it should be a separate pull / issue. Personally, I'm not that interested in the sum of lines as there are outliers that can completely skew the total. Maybe the sum of the `log()` of lines changed :-)

File: frontend/src/pull-card/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielbeardsley) [reply frontend/src/pull-card/index.tsx]: Nah, it seems even better to have them the same to me.

File: frontend/test/pull-card-demo.tsx (hunk 2 above)
Comment (jordycosta) [reply frontend/test/pull-card-demo.tsx:69]: Example added via https://github.com/iFixit/pulldasher/pull/365/commits/6d1a2f1435bcd367bbc7b14bab36cd26f0af2590 & https://github.com/iFixit/pulldasher/pull/365/commits/7aa012e49b36a6b4e261c374d9bedee3fbfd3a8d

**Comment by danielbeardsley** [comment]
CR :+1: 

Though I think the font is a little big (maybe match the pull title text?) and I'm not a fan of the grey background:
image

**Comment by jordycosta** [comment]
@danielbeardsley 

Screenshot 2023-03-23 115614

Here's using _aliceblue_ as the background, there are not many eye-friendly colors with the green/red text
https://html-color.codes/color-names/aliceblue

We can also just make it transparent, which will definitely be needed for dark mode

**Comment by danielbeardsley** [comment]
> We can also just make it transparent 

I'd opt for that since all the other pieces of the card are transparent. If the green is too light, try it a little darker :-)

**Comment by danielbeardsley** [comment]
CR :+1: 

I noticed that github no longer users a monspace font FYI:
image

**Comment by jordycosta** [comment]
I slightly modified the font & font-size to what I think looks a bit better than how it was

Before/After:

b4 after

File: frontend/src/pull-card/index.tsx [hunk 3]
```diff
@@ -103,6 +105,22 @@ export const PullCard = memo(function PullCard({
           />
           <Flags pull={pull} />
         </SigsAndFlags>
+        {showLinesChanged &&
+        <Code backgroundColor="transparent">
+          <HStack>
+            <Text fontFamily={'Verdana'}
+              title={pull.additions == 1 ? `${pull.additions} addition` : `${pull.additions} additions`}
+              fontSize={"12px"}
```
Comment (danielbeardsley) [inline frontend/src/pull-card/index.tsx:113]: For what it's worth, if you're setting a prop to a static string, you don't need curly braces. Curly braces say "evalute and return this javascript expression". IF your expression is just a static string you don't need the JS eval part: `fontSize="12px"`.

Comment (jordycosta) [reply frontend/src/pull-card/index.tsx:113]: Thank you, I'm getting a bit more used to using React so that is helpful

Comment (jordycosta) [reply frontend/src/pull-card/index.tsx]: Resolved via https://github.com/iFixit/pulldasher/pull/365/commits/80657a6243c41fc7568a846fd91d48205454eabb

File: frontend/test/pull-card-demo.tsx [hunk 4]
```diff
@@ -75,4 +77,8 @@ function Row({ title, pullDatas }: { title: string; pullDatas: PullData[] }) {
   );
 }
 
+Row.defaultProps = {
```
Comment (danielbeardsley) [inline frontend/test/pull-card-demo.tsx:80]: Hmm what's this? I wasn't aware react would examine properties on the function instance of a function component.

TIL! Looks like it used to be a thing but was deprecated in 2019:
https://github.com/facebook/react/pull/16210

File: frontend/test/pull-card-demo.tsx [hunk 5]
```diff
@@ -75,4 +77,8 @@ function Row({ title, pullDatas }: { title: string; pullDatas: PullData[] }) {
...
+  linesChanged: false,
```
Comment (danielbeardsley) [inline frontend/test/pull-card-demo.tsx:81]: Given that the default is falsy I don't think we need a default as null and undefined are both falsy.

**Review by danielbeardsley — APPROVED** [review]
CR :+1: 
 Looks good!

dev_block :+1:  on the deprecated `defaultProps` usage I think.

**Comment by jordycosta** [comment]
@danielbeardsley 

      <Row title="Labels" pullDatas={Labels} />
      <Row title="Draft" pullDatas={Draft} />
      <Row title="My Own" pullDatas={MyOwn} />
      <Row title="Kitchen Sink (all the things)" pullDatas={KitchenSink} linesChanged={true}/>

I was looking for a way to not append `linesChanged={false}` for each Row that doesn't use these line changes

Is there anyway to default the value so it's not required in each Row instance? otherwise I could just append that param

**Comment by jordycosta** [comment]
un_dev_block 🟢 

Removed defaultProps usage via https://github.com/iFixit/pulldasher/pull/365/commits/9cbe10d49892494082b3e2f8628f04e3f0c0bc34

**Comment by deltuh-vee** [comment]
QA 🎬 
Lines changed are now included for pull requests in the CR column.

**Comment by danielbeardsley** [comment]
> I was looking for a way to not append linesChanged={false} for each Row that doesn't use these line changes

Looks like you did just what I was gonna suggest, make the prop optional :-)

CR :100:

**Review by danielbeardsley — APPROVED** [review]
CR :+1:

**Pull request merged**